fs/io_zenfs.cc

```diff
@@ -215,6 +240,13 @@ Status ZoneFile::MergeUpdate(std::shared_ptr<ZoneFile> update, bool replace) {
   is_sparse_ = update->IsSparse();
   MetadataSynced();
 
+  /* update the linkfiles_ with the updated list */
```

Comment:
I think this is a bit helpful in understanding what is happening..

Reply:
In that case, I would suggest factoring out the code that the comment relates to in a function with a descriptive name
```c++
...

  UpdateLinkfilesWithUpdatedList(update); 

  return Status::OK();
}

void ZoneFile::UpdateLinkfilesWithUpdatedList(std::shared_ptr<ZoneFile> &update) {
  linkfiles_.clear();
  for (uint32_t i = 0; i < update->GetNrLinks(); i++) {
    linkfiles_.push_back(update->GetLinkFilename(i));
  }
}
```

Reply:
Why ? 
If keeping the comment is not good  because it may not be maintained properly in future, same holds true for making the comment into a function name, as the function could be modified in future to do more/less things.

Reply:
Because, if you move the function invocation around, the message of what it does automatically moves with the invocation and the lines of code that the comment/function name describes. If you have a comment attached to a few lines of code, it can be unclear how much of that code the comment refers to, and moving the code with the comment can be difficult. When refactoring the code at a later stage, the comment may move away from the thing it describes, and then at an even later point in time, someone may forget to update the comment, because it is no longer attached to the code that it actually describes.

It is true that the function name that describe the function content can also rot, but an inline comment that covers some following code rots much faster and is more difficult to maintain.

For deeper explanation check out the comment chapter in [Clean Code by Robert C. Martin](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) or [this writeup](https://medium.com/@bpnorlander/stop-writing-code-comments-28fef5272752) on medium.

Reply:
Trashing any comment as bad coding practice is hardly the right approach. I will remove the comment, but I don't agree that this is a universal rule.

Reply:
> Trashing any comment as bad coding practice is hardly the right approach.

Let's agree to disagree on that, for in-line comments at least. For API docs it's a different story.

Reply:
> I have also observed that people tend to update comments in and around the area that they modify and function headers generally tend to rot as more people add/change functionality of a function.

This is completely true, it happens. But I maintain that inline comments rot way faster than descriptive function names.

> Large scale refactors are frowned upon as the program matures.

We _should_ refactor our code _all the time_ to keep structure clean. If the code is difficult to refactor, we already failed, and we should try to correct that error. Refactoring is not to be avoided, it is to be encouraged. If we choose to break structure because we are afraid that we might break function in the process, then we have a serious problem (maybe we lack proper test coverage?).

> When someone has to fix a priority bug in an area that one is not familiar with, in short time, comments really help. 

If the comments are required to understand the code, we _already failed_ and we should try to fix that rather than make up for it with more comments. **Most of the time there is really no excuse for writing clean code.** If we stumble upon messy code when fixing a bug, we should clean up and leave the code in better shape than when we found it.

>  Infact, within fw engineering, we have tools that measure code/comments ratios.

If the goal of running that tool is to keep the amount of inline comments high, I am out of words. Just because an org chooses to follow such a doctrine does _not_ mean that it a wise choice.

> So IMO, not having comments in the code works only in an perfect world where everyone codes and thinks similarly, else things are bound to slowdown as we try to make sense of the code.

Not having comments work fine if people write their code so that intent is clear and code is easy to read. We don't have to think in a similar way for that to be effective. Most of the time inline comments is an attempt to fix our failure to express ourselves in code. I really like the following quote from Clean Code. It is a bit harsh, but I think it explains very well why we write inline comments:

> The proper use of comments is to compensate for our failure to express ourself in code. Note that I used the word failure. I meant it. Comments are always failures.

I totally respect that people have different opinions on the matter. Also, the viewpoints that I try to convey here are not viewpoints that I invented myself. It is viewpoints that I picked up by studying other peoples opinions on the subject. I find that the "write clean code and drop the comments" camp has better arguments.

For instance take this example from the blog post I posted in an earlier comment: ([source](https://gist.github.com/norMNfan/73e8b31505c0221bfcb70b1481139874#file-wronglevelofabstraction-java))
```java
// This function calculates prices, compares to sales
// promotions, checks if prices are valid, then 
// send an email of promotion to user
public void doSomeThings() {

  // Calculate prices
  ...
    ...
    ...
  
  // Compare calculated prices with sales promotions
  ...
    ...
    ...
  
  // Check if calculated prices are valid
  ...
    ...
    ...
  
  // Send promotions to users
  ...
    ...
    ...
}
```
vs this refactor: ([source](https://gist.github.com/norMNfan/9806c3b382dee9cb9d8cd9601f5e0606#file-wronglevelofabstractionrefactored-java))
```java
public void sendPromotionEmailToUsers() {
  calculatePrices();
  compareCalculatedPricesWithSalesPromotions();
  checkIfCalculatedPricesAreValid();
  sendPromotionEmail();
}
```
I think it is very clear that the refactor is easier to read and understand. 

The same argument holds for the code this comment was originally about. It is easier to read if it is properly decomposed with proper naming. Or do  you think this is not the case?